cli/replicasets/setup.go

```diff
@@ -0,0 +1,287 @@
+	}
+
+	if controlInstanceName == "" {
+		// get first instance of the first configured replicaset
+		if len(*replicasetsList) > 0 {
+			controlInstanceName = (*replicasetsList)[0].InstanceNames[0]
+		}
+	}
+
+	consoleSockPath := project.GetInstanceConsoleSock(ctx, controlInstanceName)
+	conn, err := common.ConnectToTarantoolSocket(consoleSockPath)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to Tarantool instance: %s", err)
+	}
+
+	log.Debugf("Connected to %s", consoleSockPath)
+
+	topologyReplicasets, err := getTopologyReplicasets(conn)
+	if err != nil {
+		return fmt.Errorf("Failed to get current topology replicasets: %s", err)
+	}
+
+	var errors []error
+
+	log.Debugf("Setup replicasets")
+
+	for _, replicasetConf := range *replicasetsList {
+		res := common.Result{
+			ID: replicasetConf.Alias,
+		}
+
+		topologyReplicaset := topologyReplicasets.GetByAlias(replicasetConf.Alias)
+
+		if topologyReplicaset == nil {
+			if err := createReplicaset(conn, replicasetConf, instancesConf); err != nil {
+				res.Status = common.ResStatusFailed
+				res.Error = err
+			} else {
+				res.Status = common.ResStatusCreated
+			}
+		} else {
+			if changed, err := updateReplicaset(conn, topologyReplicaset, replicasetConf, instancesConf); err != nil {
+				res.Status = common.ResStatusFailed
+				res.Error = err
+			} else if changed {
+				res.Status = common.ResStatusUpdated
+			} else {
+				res.Status = common.ResStatusOk
+			}
+		}
+
+		if res.Status == common.ResStatusFailed {
+			errors = append(errors, res.FormatError())
+		}
+
+		log.Infof(res.String())
+	}
+
+	if len(errors) > 0 {
+		for _, err := range errors {
+			log.Errorf("%s", err)
+		}
+		return fmt.Errorf("Failed to setup replicasets")
+	}
+
+	if ctx.Replicasets.BootstrapVshard {
+		if err := bootstrapVshard(conn); err != nil {
+			return fmt.Errorf("Failed to bootstrap vshard: %s", err)
+		}
+
+		log.Infof("Vshard is bootstrapped successfully")
+	}
+
+	log.Infof("Topology is set up successfully")
+
+	return nil
+}
+
+func getReplicasetsList(ctx *context.Ctx) (*ReplicasetsList, error) {
+	var err error
+
+	if _, err := os.Stat(ctx.Replicasets.File); err != nil {
+		return nil, fmt.Errorf("Failed to use replicasets configuration file: %s", err)
+	}
+
+	fileContentBytes, err := common.GetFileContentBytes(ctx.Replicasets.File)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read replicasets configuration file: %s", err)
+	}
+
+	var replicasetsConf ReplicasetsConf
+	if err := yaml.Unmarshal([]byte(fileContentBytes), &replicasetsConf); err != nil {
+		return nil, fmt.Errorf("Failed to parse replicasets configuration file %s: %s", ctx.Replicasets.File, err)
+	}
+
+	replicasetsList := make(ReplicasetsList, len(replicasetsConf))
+
+	i := 0
+	for replicasetAlias, replicasetConf := range replicasetsConf {
+		replicasetConf.Alias = replicasetAlias
+
+		replicasetsList[i] = replicasetConf
+		i++
+	}
+
+	// move vshard-router replicasets at the list beginning
```

Comment:
I guess such approach is a bit wrong.

You shouldn't sort array for that. You need to join ALL instances at the same time but not by one.
edit_topology accepts array of instances.

Reply:
I join replicasets one-by-one, not instances. The one reason is that in old versions of Cartridge joining more than one replicaset causes an error. The other is that it's much more simple to manage replicasets one-by-one (believe me, I've done it many times). About moving router to the beginning - it's not the best solution, but could you suggest something else to prevent vshard bootstrapping error?

Reply:
> The one reason is that in old versions of Cartridge joining more than one replicaset causes an error. 

Please show me commit/PR where it was fixed

Reply:
Completely rewrote this part, now all replicasets are managed by single `edit_topology` call.